grafana/grafana
# sql: use `resample` from grafana-plugin-sdk-go

as part of the decoupling effort, there are some files that are used by all the core sql plugins: mysql & mssql & postgres. we need to "solve" this:

we moved `resample.go` to `grafana-plugin-sdk-go` ( in  https://github.com/grafana/grafana-plugin-sdk-go/pull/950 )

and this PR modifies core-grafana to use that function from `grafana-plugin-sdk-go`.

NOTE: there was 1 important modification in the plugin-sdk's `resample` code:
originally `resample.go` adjusted the start-time to a nice aligned timestamp (see: https://github.com/grafana/grafana/blob/518d3341d58fd5d5bd4f73058bbbfa68cd0069ba/pkg/tsdb/sqleng/resample.go#L93-L94 ).
our "moved" function does not do this, so it has to be done at the caller's place (in `sql_engine.go`).

how to test:
- the CI job will run the postgres & mysql snapshot tests, these do exercise this code
- if you want to test manually, you need a sql query with the following properties:
    - format=time_series
    - use a macro that deals with fill-mode, like `$__timeGroup` in the sql query: example data and query:
          - create the tabe
              - `create table t1 ("time" timestamp with time zone, value double precision);`
              - `insert into t1 values ('2024-04-12T07:20:00', 15);`
              - `insert into t1 values ('2024-04-12T07:50:00', 55);`
              - `insert into t1 values ('2024-04-12T08:10:00', 2);`
          - query it
              - `select * from t1`
                  - this does not use the resample function. this should show 3 data-points
              - `SELECT $__timeGroup("time", '7m', 0), avg(value) as value FROM t1 GROUP BY time ORDER BY time`
                  - this should show a data-point every 7 minutes. it will show the original 3 points at roughly the original time-stamps, and in between them every 7 minues there will be a `0` data-point
